SurfaceTopography/Uniform/common.py

```diff
@@ -75,25 +173,45 @@ def derivative(topography, n, periodic=None):
         raise ValueError(
             'SurfaceTopography does not have physical size information, but '
             'this is required to be able to compute a derivative.')
-    grid_spacing = topography.pixel_size
-    heights = topography.heights()
+
+    if operator is None:
+        operator = _get_default_derivative_operator(n, topography.dim)
+
+    grid_spacing = np.array(topography.pixel_size)
     is_periodic = topography.is_periodic if periodic is None else periodic
-    if is_periodic:
-        der = np.array(
-            [(np.roll(heights, -1, axis=d) - heights) / grid_spacing[d]
-             for d in range(len(heights.shape))])
-        # Apply derivative into each direction multiple times
-        for i in range(n - 1):
-            der = np.array(
-                [(np.roll(der[d], -1, axis=d) - der[d]) / grid_spacing[d]
-                 for d in range(len(heights.shape))])
-    else:
-        der = np.array([np.diff(heights, n=n, axis=d) / grid_spacing[d] ** n
-                        for d in range(len(heights.shape))])
-    if der.shape[0] == 1:
-        return der[0]
+
+    # Return FFT object (this will only be initialized once and reused in subsequent calls)
+    fft = topography.make_fft()
+
+    # These fields are reused when this function is called multiple times
+    real_field = fft.fetch_or_register_real_space_field('real_temporary', 1)
+    fourier_field = fft.fetch_or_register_fourier_space_field('complex_temporary', 1)
+    np.array(real_field, copy=False)[...] = topography.heights()
+    fft.fft(real_field, fourier_field)
+
+    # Apply mask function
+    if mask_function is not None:
+        np.array(fourier_field, copy=False)[...] *= mask_function((fft.fftfreq.T / grid_spacing).T)
+
+    # Apply derivative operator in Fourier space
+    if isinstance(operator, tuple):
+        fourier_array = np.array(fourier_field, copy=False)
+        fourier_copy = fourier_array.copy()
+        der = []
+        for i, op in enumerate(operator):
+            fourier_array[...] = fourier_copy * op.fourier(fft.fftfreq)
+            fft.ifft(fourier_field, real_field)
+            _der = np.array(real_field, copy=False) * fft.normalisation / grid_spacing[i] ** n
```

Comment:
The operator encodes a certain stencil. It does not know which derivative that stencil corresponds to. We could store that information, but this would be 'meta'-information that is not required anywhere else.

Reply:
Ok

Reply:
We could extract the order of the derivative from the behavior at small q... let me try to implement this...